### ISSUE TITLE
Bytes EFBBBF (BOM) randomly disappears from middle of binary content

### ISSUE THREAD

**Issue by adcy** [Http-Multipart-Data-Parser/Http-Multipart-Data-Parser#64]
Hello!
Here is very foolish bug when BOM marker cut out from middle of content. Problem in reading input stream into buffer. Raw stream read by chunks and every chunk examined on start with BOM marker (EFBBBF). Suddenly this sequence of bytes encounters not only as BOM marker, especially if transmitted large binary files. At current if accidentally some chunk started with this three bytes even it is not the BOM, they cut out. And as result received content is not valid.

**Comment by kotVmeshke** [comment]
Hi all, I bumped with the same issue during uploading large binary files by chunks. Sometimes happens another issue - parser writes some extra bytes to output stream. I dont know why yet.

@andreycherkashin , it might be you found some workaround for this bug?

**Comment by andreycherkashin** [comment]
@kotVmeshke yes i found. Used [Microsoft.AspNetCore.WebUtilities](https://www.nuget.org/packages/Microsoft.AspNetCore.WebUtilities/2.1.1) nuget package instead.

**Comment by kotVmeshke** [comment]
@andreycherkashin thank you a lot for the quick answer! Our project based on the old good .Net 4.5 so I cannot  use this package. I found another approach yesterday https://stackoverflow.com/questions/7460088/reading-file-input-from-a-multipart-form-data-post/14514351/#21689347. It works fine.

**Comment by kalpasrl** [comment]
Hello,
in one of our apps we have the same problem.

When receiving a large file, with the chunk set to default (4096 bytes), if one of the chunks randomly starts with the EFBBBF bytes sequence, a corrupt file is produced at the end.

Is there a solution to this bug?

Thank you

**Comment by Jericho** [comment]
Can you provide the attachment that is causing this issue so we can attempt to reproduce?

**Comment by Jericho** [comment]
I wrote a unit test where I simulate a file with BOM at the beginning of a chunk. My test passes successfully. Either there's something specific to the attachment you are dealing with or there a nuance to this scenario that my test is not capturing properly.

@kalpasrl  Can you share the file that causing the problem for you? If not, can you help me reproduce this problem? I am afraid I won't be able to research this issue any further unless I get additional details.

**Comment by kalpasrl** [comment]
Thank you for your interest.

It is difficult for me to provide the file because it is for our client's confidential use.
I will keep you updated if there are any news about it.

**Comment by Jericho** [comment]
I understand the document nature of your client's document.

In that case, can you fork this repository and run it in debug mode? Maybe you can figure out where the problem is?

**Comment by Jericho** [comment]
I wrote [another unit test](https://github.com/Http-Multipart-Data-Parser/Http-Multipart-Data-Parser/commit/7ba0578375e62539330f108c222b5c87a9a9a22d) to demonstrate that we can read BOM at a random position in a stream along with encoded content. This test does not attempt to align the BOM with the beginning of a chunk which, according to preliminary info, seems to be the root of the problem (although so far unproven). I simply wanted to prove that we can read some of the most common BOM markers.

**Comment by Jericho** [comment]
Assuming you have a copy of your client's confidential document somewhere on your hard drive, here's a unit test that will read the content of this file, generate the "payload" that would normally be POSTed, parse the payload and finally assert that the content of the parsed file matches the content of the original file. 

```csharp
public class FileFromDisk
{
	[Fact]
	public void SingleFileTest()
	{
		// This is the path of the file we want to process
		var filePath = @"C:\Path\To\Your\File\ClientConfidential.pdf";

		// Create the payload
		var originalContent = File.ReadAllBytes(filePath);
		var stream = new MemoryStream();
		var writer = new StreamWriter(stream, Encoding.UTF8);
		writer.WriteLine("--boundry");
		writer.WriteLine($"Content-Disposition: form-data; name=\"myFile\"; filename=\"{Path.GetFileName(filePath)}\";");
		writer.WriteLine("Content-Type: application/octet-stream");
		writer.WriteLine("");
		writer.WriteLine(Convert.ToBase64String(originalContent));
		writer.WriteLine("--boundry--");
		writer.Flush();
		stream.Position = 0;

		// Parse the payload
		var parser = MultipartFormDataParser.Parse(stream, "boundry", Encoding.UTF8);

		// Get the content of the parsed attachment
		var rawParsedContent = new byte[parser.Files[0].Data.Length];
		parser.Files[0].Data.Read(rawParsedContent, 0, rawParsedContent.Length);
		var parsedContent = Convert.FromBase64String(Encoding.UTF8.GetString(rawParsedContent));

		// Generate hashes to compare the original content with the parsed content
		var hash = SHA1.Create();
		var originalContentHash = ConvertBytesToHex(hash.ComputeHash(originalContent));
		var parsedContentHash = ConvertBytesToHex(hash.ComputeHash(parsedContent));

		// Assert that the content of the parsed file matches the original content
		Assert.Equal(originalContentHash, parsedContentHash);
	}

	private static string ConvertBytesToHex(byte[] bytes)
	{
		var sb = new StringBuilder();
		for (var i = 0; i < bytes.Length; i++)
		{
			sb.Append(bytes[i].ToString("x"));
		}

		return sb.ToString();
	}
}
```

Let me know if this helps.

**Comment by XVII** [comment]
Running into this issue as well, and unfortunately with a file I cannot share. 

I looked at the raw request in 4096 chunks (and the individual multipart/form-data in 4096 chunks) but couldn't see any correlation. However, setting the `binaryBufferSize` to something other than 4096 lets this work.

This only difference between the working and non-working payloads are the BOM `EF BB BF` as part of 
```
<</Length 237/Type/Metadata/Subtype/XML>>stream
<?xpacket begin="ï»¿" id="
```
So it's accidentally being stripped out by the multipart/form-data library.

**Comment by Jericho** [comment]
@XVII were you able to use the C# code I provided in October 2022 along with your `BlankPdf.pdf` to reproduce the problem? I just tried and I can't reproduce.

I even improved the code to repeatedly parse the PDF with various buffer sizes and still can't reproduce:

```csharp
[Fact]
public void SingleFileTest()
{
	// This is the path of the file we want to process
	var filePath = @"C:\Temp\BlankPdf.pdf";

	// Create the payload
	var originalContent = File.ReadAllBytes(filePath);
	var stream = new MemoryStream();
	var writer = new StreamWriter(stream, Encoding.UTF8);
	var boundry = "--boundry";
	var endBoundry = boundry + "--";
	var minBufferSize = endBoundry.Length;

	writer.WriteLine(boundry);
	writer.WriteLine($"Content-Disposition: form-data; name=\"myFile\"; filename=\"{Path.GetFileName(filePath)}\";");
	writer.WriteLine("Content-Type: application/octet-stream");
	writer.WriteLine("");
	writer.WriteLine(Convert.ToBase64String(originalContent));
	writer.WriteLine(endBoundry);
	writer.Flush();

	for (int bufferSize = minBufferSize; bufferSize < 5000; bufferSize++)
	{
		try
		{
			// Reposition to the begining of the stream for each buffer size test
			stream.Position = 0;

			// Parse the payload with the specified buffer size
			var parser = MultipartFormDataParser.Parse(stream, binaryBufferSize: bufferSize);

			// Get the content of the parsed attachment
			var rawParsedContent = new byte[parser.Files[0].Data.Length];
			parser.Files[0].Data.Read(rawParsedContent, 0, rawParsedContent.Length);
			var parsedContent = Convert.FromBase64String(Encoding.UTF8.GetString(rawParsedContent));

			// Generate hashes to compare the original content with the parsed content
			var hash = SHA1.Create();
			var originalContentHash = ConvertBytesToHex(hash.ComputeHash(originalContent));
			var parsedContentHash = ConvertBytesToHex(hash.ComputeHash(parsedContent));

			// Assert that the content of the parsed file matches the original content
			Assert.Equal(originalContentHash, parsedContentHash);
		}
		catch (Exception ex)
		{
			Assert.Fail($"Failed with buffer size {bufferSize}: {ex.Message}");
		}
	}

	Debug.WriteLine("We successfully parsed the file with different buffer sizes.");
}
```

**Comment by XVII** [comment]
Hi @Jericho , thanks for the info -- I was able to produce a fuzzing example that causes the issues at iterations 2908 and 7004.

https://gist.github.com/XVII/af0e4cca013261f149d1b7ccf6b8f2fe

I've attached two example files that fail:

[ParsedInjectedPdf_2908.pdf](https://github.com/user-attachments/files/21959614/ParsedInjectedPdf_2908.pdf)
[InjectedPdf_2908.pdf](https://github.com/user-attachments/files/21959615/InjectedPdf_2908.pdf)

Also, we inject the raw binary/bytes into the HTTP request instead of converting it to Base64.

**Comment by XVII** [comment]
It might be the way the ReadLine is interpreting hte next chunk?
Looks like these functions don't keep the BOM mark which if it happens to be at the start of a chunk gets consumed.

https://stackoverflow.com/a/11701560/168457

There's discussion about checking the preamble (`Encoding.GetPreamble`) though you could argue there should be no interpretation of the BOM because it's not a string we're looking for but raw bytes.

**Comment by Jericho** [comment]
> There's discussion about checking the preamble (Encoding.GetPreamble) though you could argue there should be no interpretation of the BOM because it's not a string we're looking for but raw bytes.

I agree that there should be no interpretation of BOM when the data is binary (such as when we are parsing the content of a file). However, we do want the BOM to be eliminated when the data is a string.

The current logic eliminates the BOM in all circumstances, which is wrong. I am working on improving this logic to be more judicious and to eliminate the BOM only when appropriate.

Oh and another good news: I figured out that the unit test I previously wrote in an effort to reproduce the problem was not inserting the BOM precisely at the beginning of a chunk which explains why I couldn't reproduce. I have fixed my test and I am now able to reproduce. This great and will make it easy to validate when I have a solution.

**Comment by XVII** [comment]
> I'm making slow but steady progress due to juggling many priorities in addition to life itself, so please be patient with me. It might take me a while but I'm not going to give up until I figure this out.

No rush! I'm just happy we finally got some reproducible since 2018. Nothing worse than trying to fix a bug you can't replicate :)

**Comment by Jericho** [comment]
I have published a beta package to my personal NuGet feed ([instructions](https://github.com/Jericho/Jericho/blob/main/MyNugetFeed.md)) with a fix for this bug.  If anybody has a few minutes and would like to test and give me some feedback, that would be greatly appreciated.

**Comment by XVII** [comment]
The UTF-8 BOM is optional, which makes it tricky. I don't think we can force people to provide a BOM.

The note on [UTF8Encoding.Preamble](https://learn.microsoft.com/en-us/dotnet/api/system.text.utf8encoding.preamble?view=net-9.0) says "supplying a BOM at the beginning of an appropriate byte stream is the developer's responsibility.". Not sure which developer :)

**Comment by Jericho** [comment]
> I personally feel that there should be no BOM interpretation

Trust me: I would also prefer not to have to deal with this headache but the reality is that we have to. Let me give you a quick oversimplification of the situation:

When we receive encoded data, the very first line will look something like this:
```
0xef 0xbb 0xbf 0x45 0x45 0x98 0x111 0x117 0x110 0x100 0x97 0x114 0x121
```
When this binary data is converted to a string, it yields the following: `{BOM character}--boundary`

You would subsequently expect the following `if` condition to evaluate true but unfortunately it evaluates false due to the presence of the BOM character:
```csharp
if (line == boundary)
```

And the fact that it evaluates false completely throws off our parsing logic.

That's the whole reason why we need to remove the BOM if we find it in the data.

The currently logic removes the BOM when it ingests each 4k chunk of binary data. This works fine if the entire form is smaller than 4k or if the BOM is not precisely located at the beginning of a "chunk" of data. However this logic falls apart when the multipart form contains a binary file such as a PDF (which increases the possibility that it exceeds the 4k chunk size) and it so happens that a BOM is present at the 4k+1 position, or at the 8k+1 position, or the 12k+1 position, and so forth. This is exactly the scenario you experienced.

The new logic I am implementing is only filtering the BOM from the first chunk when ingesting data. I think this was the intent all along.

> I don't know how you'd ever handle the situations of "mixed" encodings predictably

I think I came to terms with the fact that the onus is on the developer to specify the correct encoding and our parser will behave in unpredictable ways if they don't.


Having said all this, have you had an opportunity to try the beta package I published? (FYI: I published a second beta package a few minutes ago).

**Comment by XVII** [comment]
We're doing some testing first thing next week to see if it helps resolve it via beta, thanks!

**Comment by XVII** [comment]
@Jericho , all good with `9.1.0-file-chunk-starts-with-BOM.1-16` (y)

**Comment by Jericho** [comment]
Thank you for confirming. I'm going to start preparing a new release
